ambry-cloud/src/main/java/com/github/ambry/cloud/CloudBlobStore.java

```diff
@@ -381,24 +382,52 @@ public void delete(List<MessageInfo> infos) throws StoreException {
    * Delete the specified blob if needed depending on the cache state.
    * @param blobId the blob to delete
    * @param deletionTime the deletion time
+   * @param lifeVersion life version of the blob.
    * @return whether the deletion was performed
    * @throws CloudStorageException
    */
-  private boolean deleteIfNeeded(BlobId blobId, long deletionTime) throws CloudStorageException {
+  private boolean deleteIfNeeded(BlobId blobId, long deletionTime, short lifeVersion) throws CloudStorageException {
     String blobKey = blobId.getID();
     // Note: always check cache before operation attempt, since this could be a retry after a CONFLICT error,
     // in which case the cache may have been updated by another thread.
-    if (!checkCacheState(blobKey, BlobState.DELETED)) {
-      boolean deleted = cloudDestination.deleteBlob(blobId, deletionTime);
-      addToCache(blobKey, BlobState.DELETED);
+    if (!checkCacheState(blobKey, lifeVersion, BlobState.DELETED)) {
+      boolean deleted = cloudDestination.deleteBlob(blobId, deletionTime, lifeVersion);
+      addToCache(blobKey, lifeVersion, BlobState.DELETED);
       return deleted;
     }
     return false;
   }
 
   @Override
   public short undelete(MessageInfo info) throws StoreException {
-    throw new UnsupportedOperationException("Undelete not supported in cloud store");
+    checkStarted();
+    try {
+      return requestAgent.doWithRetries(() -> undeleteIfNeeded((BlobId) info.getStoreKey(), info.getLifeVersion()),
+          "Undelete", partitionId.toPathString());
+    } catch (CloudStorageException cex) {
+      StoreErrorCodes errorCode =
+          (cex.getStatusCode() == STATUS_NOT_FOUND) ? StoreErrorCodes.ID_Not_Found : StoreErrorCodes.IOError;
+      throw new StoreException(cex, errorCode);
+    }
+  }
+
+  /**
+   * Undelete the specified blob if needed depending on the cache state.
+   * @param blobId the blob to delete.
+   * @param lifeVersion life version of the deleted blob.
+   * @return final updated life version of the blob.
+   * @throws CloudStorageException in case any exception happens during undelete.
+   */
+  private short undeleteIfNeeded(BlobId blobId, short lifeVersion) throws CloudStorageException {
+    // TODO: Currently this is implemented for undeletes via replication only for DR.
+    // See note in deleteIfNeeded.
+    if (!checkCacheState(blobId.getID(), lifeVersion, BlobState.CREATED)) {
+      short newLifeVersion = cloudDestination.undeleteBlob(blobId, lifeVersion);
+      addToCache(blobId.getID(), newLifeVersion, BlobState.CREATED);
```

Comment:
What if the blob went from state TTL_UPDATED to DELETED before getting undeleted.  Does it matter that we are switching it back to CREATED?  It seems okay, just checking.